### PR TITLE
simplify fetchers

### PR DESCRIPTION
**Pull request by nonsense**
This PR is simplifying the network fetchers functionality and mostly addressing the issues described at https://github.com/ethersphere/go-ethereum/issues/1309

0. The two new functions introduced are:
```
	NetStore.RemoteGet = delivery.RequestFromPeers
	NetStore.RemoteFetch
```

You probably want to start reviewing this PR by checking them out and checking where they are used.

1. `swarm/network/fetcher.go` has been removed and replaced with `singleflight.Group`. Every time the `Netstore.Get` looks for chunk that is not in our LocalStore, it uses the `singleflight.Group` in order to suppress duplicate retrievals for the same chunk.

Each in-flight request for a chunk adds a `FetcherItem` value to a map, so that the `NetStore.Get` receives a signal that the chunk it requested has been delivered. The signal that a chunk is delivered and stored to the LocalStore is the closing of the `FetcherItem.Delivered` channel.

`NetStore.Put` is responsible for removing fetchers and closing the delivered channel. If a fetcher item has been added to the fetchers map, then we must get that chunk and there is an interested party in the delivered signal.

We need to add a timeout to the fetchers, because even though a chunk has been requested and an actor waits for it, doesn't mean it will be delivered.

2. `handleOfferedHashesMsg` uses the `Netstore.Has` to determine if it needs a chunk (while syncing). If a chunk is needed, then a `FetcherItem` is also created, so that we keep track while the current batch is being delivered.

It is possible for a chunk to be both requested through a retrieve request and delivered through syncing independently, but there would always be only one `FetcherItem` and only one `Delivered` channel for it so that interested parties are notified when the chunk is delivered and stored.

3. `// * first it tries to request explicitly from peers that are known to have offered the chunk` - this part of the functionality of `RequestFromPeers` has been removed. We no longer control the flow through values in the `context.Context`.

4. `SyncChunkStore` has been removed.

5. `SkipCheck` functionality is removed (or about to be removed).

6. `hopCount` is not implemented correctly right now - should be fixed and/or investigated.

7. `RequestTimeout` has been split into `RequestTimeout` and `FailedPeerSkipDelay` as these have different meanings.

8. Priority queue has been removed and all calls to `peer.SendPriority` have been replaced with `peer.Send`.

9. `NetStore` moved to `network` package, and no longer adhering to `ChunkStore` interface.

10. Added `LNetStore` (come up with better name) - a `NetStore` wrapper needed for the `LazyChunkReader`.

11. Solved bug in `OfferedHashes` which is blocking the event loop for handling incoming messages on the TCP socket with a peer.

12. Solved bug in `OfferedHashes` where we request the same chunk from multiple peers via the `OfferedHashes/WantedHashes` protocol.

13. Solved multiple bugs/improvements on `smoke tests`.

---

TODO:
- [ ] fix `.go` files that were renamed to `.god` so that they are not compiled. fix tests.
- [x] fix `SkipCheck`
- [x] fix `hopCount`
- [x] update `ChunkStore` interface or just add methods to `NetStore`. We probably need `GetWithCallback` (need better name), so that we atomically try to get a chunk, and if it is not there, we receive back a channel to notify us when it is delivered.
- [ ] RequestFromPeers quit channel - notify that a peer was disconnected and we don't have to wait for `1 * time.Second (SearchTimeout)` - we can immediately call next peer.
- [ ] fix log levels - currently chunk not found results in errors, which is not right.
- [x] fix mess with `RetrieveRequestMsg` and `Request` within `NetStore`

### CODE REVIEW

File: swarm/network/stream/syncer.go [hunk 1]
```diff
@@ -126,76 +143,54 @@ func (s *SwarmSyncerServer) SetNextBatch(from, to uint64) ([]byte, uint64, uint6
...
+		return NewSwarmSyncerClient(p, netStore, NewStream("SYNC", t, live))
 	})
 }
 
-// NeedData
 func (s *SwarmSyncerClient) NeedData(ctx context.Context, key []byte) (wait func(context.Context) error) {
-	return s.store.FetchFunc(ctx, key)
+	start := time.Now()
+
+	has, fi := s.store.HasWithCallback(ctx, key, "syncer")
```
Comment (zelig) [inline swarm/network/stream/syncer.go]: here we need to distinguish between
- if we has the chunk - then return nil function, wantedhashes bit will be false
- if we has no chunk - no fetcher - the newly created fetcher wait function is returned, wantedhashes bit will be true
- if we have no chunk but got the fetcher already, then the peer should be sent as source offering the chunk ,  in this case the fi based wait function should be returned and waited on before batch is complete, but the wantedhashes bit should be false.
note that this case is important, currently all synced chunks seem to be both requested AND Synced with two deliveries.
if we just rely on syncing, required chunks may come with huge delay. This latter scenario may be masked if no syncing and retrieval does not overlap but also if all synced chunks are needed for retrieval  
a source channel on `fi` is a very simple solution to have this functionality again.

Comment (zelig) [reply swarm/network/stream/syncer.go]: to handle the same scenario and allow synced chunks that hit a request to be escalated to a targeted request and 'jump queue ' , we may need to reinstate the priority queue. will see

Comment (nonsense) [reply swarm/network/stream/syncer.go]: @zelig we have timers that measure that. Once we solve the underlying reason for the timeouts and we have a stable syncing/retrieval process without thousands of timeouts for a single download request, we can make the decision whether we need to `jump queue`.

**Comment by nonsense** [comment]
@janos @zelig I changed the base branch to our release candidate branch `swarm-rather-stable`, and will start making smaller PRs, so that I make it easier to review.

Will keep this PR open because it has a meaningful description and encompasses almost all bugs and problems that this refactor is addressing. (from 0 to 13).

**Pull request closed without merging**